horovod/common/controller.h

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef HOROVOD_CONTROL_MANAGER_H
+#define HOROVOD_CONTROL_MANAGER_H
+
+#include <iostream>
+#include <queue>
+#include <vector>
+
+#include "half.h"
+#include "parameter_manager.h"
+#include "response_cache.h"
+#include "stall_inspector.h"
+#include "tensor_queue.h"
+#include "timeline.h"
+
+#if HAVE_GLOO
+#include "gloo_context.h"
+#endif
+
+namespace horovod {
+namespace common {
+
+// Forward declaration
+class HorovodGlobalState;
+
+using MessageTable = std::unordered_map<std::string, std::vector<Request>>;
+
+class Controller : public std::enable_shared_from_this<Controller> {
+public:
+  Controller(ResponseCache& response_cache,
+             TensorQueue& tensor_queue, Timeline& timeline,
+             ParameterManager& parameter_manager);
+
+  Controller(const Controller&) = delete;
+  // Functions must be overridden by concrete controller
+  virtual void Initialize() = 0;
+
+  virtual int GetTypeSize(DataType dtype) = 0;
+
+  virtual void CrossRankBitwiseAnd(std::vector<long long>& bitvector,
+                                   int count) = 0;
+
+  virtual void CrossRankBitwiseOr(std::vector<long long>& bitvector,
+                                  int count) = 0;
+
+  virtual void SynchronizeParameters() = 0;
+
+  // Concrete controller functions
+  ResponseList ComputeResponseList(std::atomic_bool& shut_down);
+
+  int64_t TensorFusionThresholdBytes();
+
+  void SetRank(const int* ranks, int nrank) {
+    for (auto i = 0; i < nrank; ++i) {
```

Comment:
Since this is public and can be called more than once, you should first have it clean the existing `rank_` to replace the contents with the new ranks.

Reply:
Fixed.